File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.cpp
class Solution {
public:

    int sumSubarrayMins(vector<int>& arr) {

        int n = arr.size();

        vector<int> prev(n);
        vector<int> next(n);

        stack<int> st;

        // Previous Smaller

        for(int i = 0; i < n; i++) {

            while(!st.empty() &&
                  arr[st.top()] > arr[i]) {

                st.pop();
            }

            prev[i] = st.empty() ? -1 : st.top();

            st.push(i);
        }

        while(!st.empty())
            st.pop();

        // Next Smaller

        for(int i = n - 1; i >= 0; i--) {

            while(!st.empty() &&
                  arr[st.top()] >= arr[i]) {

                st.pop();
            }

            next[i] = st.empty() ? n : st.top();

            st.push(i);
        }

        long long ans = 0;
        long long MOD = 1e9 + 7;

        for(int i = 0; i < n; i++) {

            long long left = i - prev[i];
            long long right = next[i] - i;

            ans = (ans +
                   (1LL * arr[i] * left * right) % MOD)
                   % MOD;
        }

        return ans;
    }
};

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        MOD = 10**9 + 7
        n = len(arr)
        prev_smaller = [-1] * n
        next_smaller = [n] * n
        stack = []

        # Previous Smaller Element (strictly smaller)
        for i in range(n):
            while stack and arr[stack[-1]] >= arr[i]:
                stack.pop()
            prev_smaller[i] = stack[-1] if stack else -1
            stack.append(i)

        stack.clear()

        # Next Smaller-or-Equal Element
        for i in range(n - 1, -1, -1):
            while stack and arr[stack[-1]] > arr[i]:
                stack.pop()
            next_smaller[i] = stack[-1] if stack else n
            stack.append(i)

        result = 0
        for i in range(n):
            left = i - prev_smaller[i]
            right = next_smaller[i] - i
            result = (result + arr[i] * left * right) % MOD
        return result